# Analysis of scRNA-seq and ATAC-seq data from ultra-high-risk neuroblastoma

## Folders

* `data_generated`: output files generated by the scripts in this repository. The following datasets are available:

  - `3_datasets_sc`: Three datasets integrated with SCTransform; small datasets that can be used for testing purposes
  - `all_datasets_sc_old`: Initial attempt at integrating all datasets with SCTransform; one sample was duplicated (with different read lengths)
  - `all_datasets_sc`: All (13) datasets integrated with SCTransform

* `data_raw`: Cell Ranger output downloaded from [here](https://biomedical-sequencing.at/projects/BSA_0407_STM_Neuroblastoma_2ba0210fb73d412397728e8a97a3e423). Subfolders ending in `_transcriptome` and `_ATAC` contain scRNA-seq and ATAC-seq data, respectively. `metadata` contains various additional information (as provided by our collaborators). Notably, `sample_groups.csv` contains mappings from samples to neuroblastoma groups.

* `doc`: project documentation

* `plots`: generated plots


## Main workflow

* [integrate_samples.R] - integrate selected scRNA-seq samples; invoke via [sge_job_integrate.sh]
* [cell_types_singler.R] - perform cell type classification via singler; invoke via [sge_job_singler.sh]
* [extract_seurat_data.R] - extract important data from (large) Seurat object to (smaller) CSV files
* [differential_expression.R] - find conserved markers; invoke via [sge_job_diffexp.sh]
* [summary_integration.R] - analysis of integrated data

## Other analyses

* [compare_reads.Rmd] - compare two samples that differ in read length (50 vs 75 bp)
* [download_data.R] - download selected raw data
* [summary_cellranger_metrics.R] - generate summary plots of Cell Ranger QC metrics
* [summary_cellranger_tsne.R] - generate overview of tSNEs calculated by Cell Ranger
+ [test_workflow.R] - template of the overall Seurat workflow (can be eventually deleted!)
* [tutorial_monocle] - misc experiments with monocle (can be eventually deleted!)