# Analysis of scRNA-seq and scATAC-seq data from bone marrow cells of ultra-high-risk neuroblastoma patients

## Folders

* `data_generated`: output files generated by the scripts in this repository.
* `data_raw`: cellranger output downloaded from [here](https://biomedical-sequencing.at/projects/BSA_0407_STM_Neuroblastoma_2ba0210fb73d412397728e8a97a3e423). Subfolders ending in `_transcriptome` and `_ATAC` contain scRNA-seq and ATAC-seq data, respectively.
* `metadata`: additional information; e.g., `sample_groups.csv` contains mappings from samples to neuroblastoma groups.
* `doc`: project documentation
* `literature`: relevant publications
* `plots`: generated plots



## Main workflow

* [merge_rna.R](merge_rna.R) -
  merge samples, perform QC filtering, ensure unique cell names;
  invoke via [sge_merge_rna.sh](sge_merge_rna.sh)
* [integrate_rna_seurat.R](integrate_rna_seurat.R) -
  integrate scRNA-seq samples with Seurat; 
  invoke via [integrate_rna_seurat.sh](sge_integrate_rna_seurat.sh)
* [extract_metadata_seurat.R](extract_metadata_seurat.R) -
  perform basic analyses and extract resulting metadata from a Seurat object,
  export individual assays
* [integrate_rna_monocle.R](integrate_rna_monocle.R) -
  integrate scRNA-seq samples with monocle, perform basic analyses and extract
  resulting metadata
* [classify_cell_types.R](classify_cell_types.R) -
  perform cell type classification via SingleR;
  invoke via [sge_classify_cell_types.sh](sge_classify_cell_types.sh)
* [detect_doublets.R](detect_doublets.R) -
  detect doublets via scds;
  invoke via [sge_detect_doublets.sh](sge_job_doublet.sh)
* [assemble_metadata.R](assemble_metadata.R) -
  generate one CSV and RDS file with all metadata
* [analyse_rna.R](analyse_rna.R) -
  analysis of integrated data
* [plot_markers.R](plot_markers.R) -
  plot canonical cell type and NB markers
* [find_conserved_markers.R](find_conserved_markers.R) -
  find conserved markers
* [correct_ambiance.R](correct_ambiance.R) –
  remove cell-free RNA contamination
* [dge_muscat.R](dge_muscat.R) -
  differential expression via muscat



## Other analyses

* [common_functions.R](common_functions.R) -
  functions used throughout the project
* [compare_reads.Rmd](compare_reads.Rmd) -
  compare two samples that differ in read length (50 vs 75 bp)
* [download_data.R](download_data.R) -
  download raw data from the BSF
* [plot_dependencies.R](plot_dependencies.R) –
  plot the dependency graph
* [plot_qc_metrics.R](plot_qc_metrics.R) -
  plot QC metrics for cell filtering (i.e., % of mitochondrial genes etc.)
* [summarize_cellranger_metrics.R](summarize_cellranger_metrics.R) -
  generate summary plots of Cell Ranger QC metrics
* [summarize_cellranger_tsne.R](summarize_cellranger_tsne.R) -
  generate overview of tSNEs calculated by Cell Ranger


## Currently unused

* [make_subclusters.R](make_subclusters.R) -
  perform subclustering
* [process_droplets.R](process_droplets.R) –
  droplet calling and QC filtering