# Analysis of scRNA-seq and scATAC-seq data from bone marrow cells of ultra-high-risk neuroblastoma patients

## Folders

- `data_generated`: output files generated by the scripts in this repository.
- `data_raw`: cellranger output downloaded from the [BSF](https://biomedical-sequencing.at/projects/BSA_0407_STM_Neuroblastoma_2ba0210fb73d412397728e8a97a3e423). Subfolders ending in `_transcriptome` and `_ATAC` contain scRNA-seq and ATAC-seq data, respectively.
- `data_wip`: data from work in progress
- `doc`: project documentation
- `literature`: relevant publications
- `metadata`: additional information; e.g., `sample_groups.csv` contains mappings from samples to neuroblastoma groups.
- `plots`: generated plots
- `rna_velocity`: RNA velocity analysis via velocyto/scvelo
  (see separate [readme](rna_velocity/README.md) for further information)
- `tables`: exported tables



## Main workflow

![workflow](plots/dependency_graph.png)

- [perform_qc.R](perform_qc.R) -
  perform QC filtering, ensure unique cell names
- [integrate_rna.R](integrate_rna.R) -
  integrate scRNA-seq samples with monocle, perform basic analyses
  and extract resulting metadata
- [correct_ambiance.R](correct_ambiance.R) –
  remove cell-free RNA contamination
- [make_subclusters.R](make_subclusters.R) -
  perform subclustering
- [classify_cell_types.R](classify_cell_types.R) -
  perform cell type classification via SingleR;
  invoke via [sge_classify_cell_types.sh](sge_classify_cell_types.sh)
- [assemble_metadata.R](assemble_metadata.R) -
  generate one CSV and RDS file with all metadata
- [analyse_rna.R](analyse_rna.R) -
  analysis of integrated data
- [plot_markers.R](plot_markers.R) -
  plot manually selected canonical cell type markers
- [find_markers.R](find_markers.R) -
  automatically detect canonical cell type markers and plot them
- [analyse_dge_pb.R](analyse_dge_pb.R) -
  analyse differential gene expression using pseudobulks
- [analyse_dge_mm.R](analyse_dge_mm.R) -
  analyse differential gene expression using mixed models
- [analyse_cnv.R](analyse_cnv.R) -
  analyse copy number variations
- [plot_cnv.R](plot_cnv.R) - 
  plot copy number variations



## Other analyses

- [common_functions.R](common_functions.R) -
  functions used throughout the project
- [compare_reads.Rmd](compare_reads.Rmd) -
  compare two samples that differ in read length (50 vs 75 bp)
- [download_data.R](download_data.R) -
  download raw data from the BSF
- [plot_dependencies.R](plot_dependencies.R) –
  plot the dependency graph, which is shown above
- [styling.R](styling.R) –
  functions for generating publication-quality figures and tables
- [summarize_cellranger_metrics.R](summarize_cellranger_metrics.R) -
  generate summary plots of Cell Ranger QC metrics
- [summarize_cellranger_tsne.R](summarize_cellranger_tsne.R) -
  generate overview of tSNEs calculated by Cell Ranger
- R scripts starting with `wip` represent work in progress
  that has not yet been merged into the main workflow
