# Summarize metrics generated by Cell Ranger
# (i.e., as stored in [sample]/metrics_summary.csv)
#
# @DEPI raw data

library(tidyverse)
library(fs)



# RNA-seq -----------------------------------------------------------------

df <- 
  dir_ls("data_raw/", recurse = TRUE, regex = "metrics_summary.csv$") %>% 
  map_dfr(read_csv, .id = "file") %>%
  rename_with(~str_glue("{.x} (%)"), !file & where(is.character)) %>% 
  mutate(across(!file & where(is.character), parse_number)) %>%
  extract(file, into = "sample", regex = "([\\w\\d_]*)_trans")

broken_labels = c(
  "Reads Mapped Antisense to Gene (%)" = "Reads Mapped\nAntisense to Gene (%)",
  "Reads Mapped Confidently to Genome (%)" = "Reads Mapped Confidently\nto Genome (%)",
  "Reads Mapped Confidently to Intergenic Regions (%)" = "Reads Mapped Confidently\nto Intergenic Regions (%)",
  "Reads Mapped Confidently to Intronic Regions (%)" = "Reads Mapped Confidently\nto Intronic Regions (%)",
  "Reads Mapped Confidently to Exonic Regions (%)" = "Reads Mapped Confidently\nto Exonic Regions (%)",
  "Reads Mapped Confidently to Transcriptome (%)" = "Reads Mapped Confidently\nto Transcriptome (%)"
)

df %>% 
  pivot_longer(!sample, names_to = "measure") %>% 
  mutate(
    measure = measure %>%
      recode(!!!broken_labels) %>% 
      as_factor()
  ) %>% 
  ggplot(aes(measure, value)) +
  geom_boxplot(width = .25, outlier.shape = NA) +
  geom_jitter(
    aes(color = sample),
    position = position_jitter(width = .1, seed = 42),
    alpha = .75
  ) +
  scale_x_discrete(NULL) +
  scale_y_continuous(NULL) +
  facet_wrap(vars(measure), scales = "free") +
  theme(strip.text = element_blank()) +
  ggtitle("Cellranger summary statistics for RNA-seq") +
  NULL

dir_create("plots/cellranger_summary")  
ggsave("plots/cellranger_summary/measures_rna.pdf",
       units = "mm", width = 297, height = 210)



# ATAC-seq ----------------------------------------------------------------

df <- 
  dir_ls("data_raw/", recurse = TRUE, regex = "/summary.csv$") %>% 
  map_dfr(read_csv, na = "None", .id = "file") %>%
  extract(file, into = "sample", regex = "([\\w\\d_]*)_ATAC")

broken_labels = c(
  "frac_fragments_overlapping_peaks" = "frac_fragments_\noverlapping_peaks",
  "frac_fragments_overlapping_targets" = "frac_fragments_\noverlapping_targets",
  "median_per_cell_unique_fragments_at_30000_RRPC" = "median_per_cell_unique_\nfragments_at_30000_RRPC",
  "median_per_cell_unique_fragments_at_50000_RRPC" = "median_per_cell_unique_\nfragments_at_50000_RRPC",
  "median_per_cell_total_library_complexity" = "median_per_cell_\ntotal_library_complexity"
)

df %>% 
  select(!`cellranger-atac_version`) %>% 
  pivot_longer(!sample, names_to = "measure") %>% 
  mutate(
    measure = measure %>%
      recode(!!!broken_labels) %>% 
      as_factor()
  ) %>% 
  ggplot(aes(measure, value)) +
  geom_boxplot(width = .25, outlier.shape = NA) +
  geom_jitter(
    aes(color = sample),
    position = position_jitter(width = .1, seed = 42),
    alpha = .75
  ) +
  scale_x_discrete(NULL) +
  scale_y_continuous(NULL) +
  facet_wrap(vars(measure), scales = "free") +
  theme(strip.text = element_blank()) +
  ggtitle("Cellranger summary statistics for ATAC-seq") +
  NULL

ggsave("plots/cellranger_summary/measures_atac.pdf",
       units = "mm", width = 297, height = 210)
